README.md
# fused-effects-gui
a haskell gui library, easy to use..


CHANGELOG.md
# Revision history for fused-effects-gui

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/MyLib.hs
{-# LANGUAGE OverloadedStrings #-}

module MyLib where

import Data.Text (Text, pack)
import Foreign.C.Types (CInt)
import SDL
import SDL.Font as SF
import SDL.Framerate
import SDL.Primitive

main :: IO ()
main = do
  initializeAll
  SF.initialize
  window <-
    createWindow
      "resize"
      WindowConfig
        { windowBorder = True,
          windowHighDPI = False,
          windowInputGrabbed = False,
          windowMode = Windowed,
          windowGraphicsContext = NoGraphicsContext,
          windowPosition = Wherever,
          windowResizable = True,
          windowInitialSize = V2 800 600,
          windowVisible = True
        }
  renderer <- createRenderer window (-1) defaultRenderer
  addEventWatch $ \ev ->
    case eventPayload ev of
      WindowSizeChangedEvent sizeChangeData ->
        putStrLn $ "eventWatch windowSizeChanged: " ++ show sizeChangeData
      _ -> return ()
  fm <- manager
  set fm 30
  font <- load "/usr/share/fonts/truetype/ubuntu/UbuntuMono-R.ttf" 20
  appLoop renderer font

blue = V4 0 0 255 255

appLoop :: Renderer -> Font -> IO ()
appLoop renderer font = do
  waitEvent >>= go
  where
    go :: Event -> IO ()
    go ev = do
      rendererDrawColor renderer $= V4 255 255 255 255
      clear renderer
      rendererDrawColor renderer $= V4 255 0 0 255
      -- drawLine renderer (P $ V2 0 0) (P $ V2 400 300)
      -- thickLine renderer 0 300 10 blue
      -- fillCircle renderer 300 20 blue
      let val = "wellcome haskell!!!!!!"
      renderFont font renderer val (P 30) blue
      present renderer
      case eventPayload ev of
        WindowSizeChangedEvent sizeChangeData -> do
          putStrLn $ "waitEvent windowSizeChanged: " ++ show sizeChangeData
          waitEvent >>= go
        KeyboardEvent keyboardEvent
          | keyboardEventKeyMotion keyboardEvent == Pressed
              && keysymKeycode (keyboardEventKeysym keyboardEvent) == KeycodeQ ->
            return ()
        MouseMotionEvent mv@(MouseMotionEventData _ _ _ p _) -> do
          let val = pack (show p)
          renderFont font renderer val (fmap fromIntegral p) blue
          present renderer
          waitEvent >>= go
        _ -> waitEvent >>= go

renderFont :: Font -> Renderer -> Text -> Point V2 CInt -> SF.Color -> IO ()
renderFont font renderer val pos color = do
  (w, h) <- size font val
  sur <- blended font color val
  texture <- createTextureFromSurface renderer sur
  copy
    renderer
    texture
    Nothing
    ( Just $
        Rectangle
          pos
          (V2 (fromIntegral w) (fromIntegral h))
    )
  freeSurface sur
  destroyTexture texture

src/Widget.hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE DeriveFoldable #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE UndecidableSuperClasses #-}

module Widget where

import Control.Algebra
import Control.Carrier.Lift
import Control.Carrier.Reader
import Control.Carrier.State.Strict
import Control.Monad
import Control.Monad.IO.Class
import Data.Dynamic
import Data.Foldable (forM_)
import Data.Kind
import Data.Text (Text, pack)
import Data.Typeable
import Data.Word (Word8)
import MyLib
import Optics
import SDL
import SDL.Font as SF
import SDL.Framerate
import SDL.Primitive

{-
Widget m

-}

data UIEnv = UIEnv
  { _renderer :: Renderer,
    _font :: Font,
    _manager :: Manager,
    _getUserEvent :: Event -> IO (Maybe Dynamic)
  }

data UIState = UIState
  { _bodyWidget :: SomeWidget,
    _focus :: [Int]
  }

type UI sig m = Has (Reader UIEnv :+: State UIState) sig m --- very cool!!!!!!

--
--
--
--  WidgetMoveChild in event
--

-- type Event' = Int

class WidgetHandler a where
  handler :: (UI sig m, MonadIO m) => [Event] -> Widget a -> m ()

--   handler :: Event' -> Widget sig m a -> m ()

type BasePositon = Point V2 Int

class WidgetRender a where
  renderSelf :: (UI sig m, MonadIO m) => BasePositon -> Widget a -> m ()

  render :: (UI sig m, MonadIO m) => BasePositon -> Widget a -> m ()
  render bp w = do
    when (_visible w) $ do
      renderer <- asks _renderer
      liftIO $ do
        rendererDrawColor renderer $= _backgroundColor w
        drawRect renderer (Just $ Rectangle (fmap fromIntegral bp) (V2 (fromIntegral (_width w)) (fromIntegral (_heigh w))))
      renderSelf bp w
      forM_ (_children w) $ \(bpc, SomeWidget w) -> render (bp + bpc) w

data Widget model = Widget
  { _width :: Int,
    _heigh :: Int,
    _model :: model,
    _backgroundColor :: V4 Word8,
    _frontColor :: V4 Word8,
    _visible :: Bool,
    _path :: [Int],
    _children :: [(BasePositon, SomeWidget)]
  }

data SomeWidget
  = forall a.
    ( WidgetRender a,
      WidgetHandler a
    ) =>
    SomeWidget (Widget a)

makeLenses ''Widget
makeLenses ''UIState
makeLenses ''UIEnv

renderSomeWidget :: (UI sig m, MonadIO m) => BasePositon -> SomeWidget -> m ()
renderSomeWidget bp (SomeWidget w) = render bp w

handlerSomeWidget :: (UI sig m, MonadIO m) => [Event] -> SomeWidget -> m ()
handlerSomeWidget es (SomeWidget w) = handler es w

children' :: Lens' SomeWidget [(BasePositon, SomeWidget)]
children' = lens (\(SomeWidget w) -> w ^. children) (\(SomeWidget w) a -> SomeWidget (w & children .~ a))

width' :: Lens' SomeWidget Int
width' = lens (\(SomeWidget w) -> w ^. width) (\(SomeWidget w) a -> SomeWidget (w & width .~ a))

height' :: Lens' SomeWidget Int
height' = lens (\(SomeWidget w) -> w ^. heigh) (\(SomeWidget w) a -> SomeWidget (w & heigh .~ a))

path' :: Lens' SomeWidget [Int]
path' = lens (\(SomeWidget w) -> w ^. path) (\(SomeWidget w) a -> SomeWidget (w & path .~ a))

-- listIdex :: Lens' [a] Int

-- model' :: (WidgetRender model, WidgetHandler model) => Lens' SomeWidget model
-- model' = lens (\(SomeWidget w) -> w ^. model) (\(SomeWidget w) a -> SomeWidget (w & model .~ a))

src/LZipper.hs
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TemplateHaskell #-}

module LZipper where

import Data.IORef
import Data.Tree
import Data.Tree.Zipper
import GHC.IOArray (newIOArray)
import Optics

t :: Tree Int
t =
  Node
    10
    [ Node 9 [],
      Node 8 [Node 81 [Node 82 []], Node 88 []],
      Node 7 []
    ]

dt = putStrLn $ drawTree $ fmap show t

-- >>> children tp
-- Loc {_content = E, _before = [], _after = [Node {rootLabel = 9, subForest = []},Node {rootLabel = 8, subForest = [Node {rootLabel = 81, subForest = [Node {rootLabel = 82, subForest = []}]},Node {rootLabel = 88, subForest = []}]},Node {rootLabel = 7, subForest = []}], _parents = [([],10,[])]}
tp = fromTree t

class Render a

class Handler a

data SomeModel = forall a b. (Render b, Handler b) => SomeModel (IORef a, Lens' a b)

instance Show SomeModel where
  show _ = "SomeModel"

data Widget' = Widget'
  { _pos :: Int,
    _size :: Int,
    _model :: SomeModel
  }
  deriving (Show)

data A = A
  { _va :: Int,
    _vb :: String
  }
  deriving (Show)

makeLenses ''A

df = A 10 "nice"

instance Render A

instance Handler A

id' :: Lens' s s
id' = lens id (\s a -> a)

mkA :: IORef A -> Widget'
mkA ref =
  Widget'
    { _pos = 1,
      _size = 2,
      _model = SomeModel (ref, id')
    }

mkW ::
  (Handler b, Render b) =>
  IORef a ->
  Lens' a b ->
  Widget'
mkW ref len =
  Widget'
    { _pos = 1,
      _size = 2,
      _model = SomeModel (ref, len)
    }

type Widget = Tree Widget'

instance Handler Int

instance Render Int

instance Handler String

instance Render String

wt :: IO ()
wt = do
  ref <- newIORef df
  let w' = mkA ref
      w =
        Node
          w'
          [ Node (mkW ref va) [],
            Node (mkW ref vb) []
          ]
  undefined


src/NewWidget.hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE DeriveFoldable #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeOperators #-}

module NewWidget where

import Control.Algebra
import Control.Carrier.Lift
import Control.Carrier.Reader
import Control.Carrier.State.Strict
import Control.Monad
import Control.Monad.IO.Class
import Data.Dynamic
import Data.Foldable (forM_)
import Data.Functor (Functor)
import Data.IORef
import Data.Kind
import Data.Text (Text, pack)
import Data.Typeable
import Data.Word (Word8)
import MyLib
import Optics
import SDL
import SDL.Font as SF
import SDL.Framerate
import SDL.Primitive
import System.IO.Unsafe

{-
Widget m

-}

data UIEnv = UIEnv
  { _renderer :: Renderer,
    _font :: Font,
    _manager :: Manager,
    _getUserEvent :: Event -> IO (Maybe Dynamic)
  }

data UIState = UIState
  { _bodyWidget :: Int, -- SomeWidget,
    _focus :: [Int]
  }

type UI sig m = Has (Reader UIEnv :+: State UIState) sig m

data Widget model = forall a.
  Widget
  { rect :: Rectangle Int,
    model :: (IORef a, Lens' a model),
    children :: [SomeWidget]
  }

instance Show model => Show (Widget model) where
  show (Widget rec (ref, m) children) =
    show rec ++ " \n"
      ++ unsafePerformIO
        ( do
            v <- readIORef ref
            return (show $ v ^. m)
        )
      ++ " \n"
      ++ show children

data SomeWidget = forall model. Show model => SomeWidget (Widget model)

instance Show SomeWidget where
  show (SomeWidget a) = show a

data A = A
  { _va :: Int,
    _vb :: String
  }
  deriving (Show)

makeLenses ''A

da = A 10 "nice"

-- >>> show $ da & id' %~ (& va .~ 10000)
-- "A {_va = 10000, _vb = \"nice\"}"
id' :: Lens' a a
id' = lens id (\a b -> b)

widgetCreate :: Show b => IORef a -> Lens' a b -> SomeWidget
widgetCreate ref len =
  SomeWidget $
    Widget
      { rect = Rectangle 30 40,
        model = (ref, len),
        children = []
      }

t :: IO ()
t = do
  refA <- newIORef (A 10 "nice")
  let w =
        Widget
          { rect = Rectangle 30 40,
            model = (refA, id'),
            children =
              [ widgetCreate refA va,
                widgetCreate refA vb
              ]
          }
  --   nw = w {parent = undefined}
  print w

data T a = T
  { val :: a,
    tl :: [T a]
  }

data C a = C
  { pv :: a,
    pl :: [T a],
    cv :: T a,
    pr :: [T a]
  }

data Zipper a = Zipper
  { _zval :: a,
    _left :: [Zipper a],
    _center :: Zipper a,
    _right :: [Zipper a]
  }

makeLenses ''Zipper

-- instance Show a => Show (Zipper a) where
--   show (Zipper ls c rs) = show ls ++ " [" ++ show c ++ "] " ++ show rs

-- leftMove :: Zipper a -> Zipper a
-- leftMove (Zipper (l : ls) c rs) = Zipper ls l (c : rs)
-- leftMove a = a

-- rightMove :: Zipper a -> Zipper a
-- rightMove (Zipper ls c (r : rs)) = Zipper (c : ls) r rs
-- rightMove a = a

-- >>> tz
-- >>> leftMove tz
-- >>> rightMove $ rightMove tz
-- [3,2,1] [4] [5,6,7]
-- [2,1] [3] [4,5,6,7]
-- [5,4,3,2,1] [6] [7]
-- tz :: Zipper Int
-- tz = Zipper [3, 2, 1] 4 [5, 6, 7]

-- type ZZ a = Zipper (Zipper a)

-- >>> tzz
-- >>> tzz & center %~ leftMove
-- >>> leftMove  tzz
-- [[3,2,1] [4] [5,6,7]] [[3,2,1] [4] [5,6,7]] [[3,2,1] [4] [5,6,7]]
-- [[3,2,1] [4] [5,6,7]] [[2,1] [3] [4,5,6,7]] [[3,2,1] [4] [5,6,7]]
-- [] [[3,2,1] [4] [5,6,7]] [[3,2,1] [4] [5,6,7],[3,2,1] [4] [5,6,7]]
-- tzz :: ZZ Int
-- tzz = Zipper [tz] tz [tz]

-- tk = tzz & center %~ leftMove

-- data Fix f a
--   = Fix (f (Fix f a))
--   | End a
--   deriving (Functor)

-- instance (Foldable f, Show a) => Show (Fix f a) where
--   show (End a) = "End " ++ show a
--   show (Fix v) = "Fix " ++ concatMap show v

-- type ZF a = Fix Zipper a

-- -- >>> tzf
-- -- Fix Fix End 0End 10End 100End 1000End 1End 2Fix End 3
-- tzf :: ZF Int
-- tzf =
--   Fix
--     ( Zipper
--         [Fix (Zipper [End 0, End 10, End 100, End 1000] (End 1) [])]
--         (End 2)
--         [Fix (Zipper [] (End 3) [])]
--     )


src/Example.hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE DeriveFoldable #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE UndecidableSuperClasses #-}

module Example where

import Control.Algebra
import Control.Carrier.Lift
import Control.Carrier.Reader
import Control.Carrier.State.Strict
import Control.Effect.Optics (use, (%=), (.=))
import Control.Monad
import Control.Monad.IO.Class
import Data.Foldable (forM_)
import Data.Kind
-- import Optics

import Data.Maybe
import Data.Text (Text, pack)
import Data.Word (Word8)
import GHC.Real (Integral)
import MyLib
import Optics (ix, (%), (^.), _1)
import SDL
import SDL.Font as SF
import SDL.Framerate
import SDL.Primitive
import Widget

data Body = Body

bodyWidget' :: Widget Body
bodyWidget' =
  Widget
    { _width = 100,
      _heigh = 100,
      _model = Body,
      _backgroundColor = 255,
      _frontColor = 90,
      _visible = True,
      _path = [],
      _children =
        [ (10, SomeWidget (modelWidget [0])),
          (110, SomeWidget (modelWidget [1]))
        ]
    }

instance WidgetRender Body where
  renderSelf bp w@Widget {..} = do
    renderer <- asks _renderer
    clear renderer

inRectangle :: (Ord a, Num a, Integral b) => Point V2 a -> Rectangle b -> Bool
inRectangle (P v0@(V2 x0 y0)) (Rectangle (P vs) size) =
  let V2 x1 y1 = fmap fromIntegral vs
      V2 x2 y2 = fmap fromIntegral (vs + size)
   in x0 > x1 && x0 < x2 && y0 > y1 && y0 < y2

-- v0 > fmap fromIntegral vs && v0 < fmap fromIntegral (vs + size)

instance WidgetHandler Body where
  handler es a = mapM_ handler1 es >> return ()
    where
      handler1 e = do
        case eventPayload e of
          -- (MouseButtonEvent (MouseButtonEventData _ Pressed _ ButtonLeft _ pos)) -> do
          --   cs <- use $ bodyWidget % children'
          --   let newmw = modelWidget [length cs]
          --   bodyWidget % children' %= ((fmap fromIntegral pos, SomeWidget newmw) :)
          -- drag model
          (MouseMotionEvent (MouseMotionEventData _ _ [ButtonLeft] pos relPos)) -> do
            ls <- use (bodyWidget % children')
            let vs = map (\(bp, sw) -> (Rectangle bp (V2 (sw ^. width') (sw ^. height')), sw ^. path')) ls
                posIn = listToMaybe $ map snd $ filter (inRectangle pos . Prelude.fst) vs
            case posIn of
              Just [i] -> do
                bodyWidget % children' % ix i % _1 %= (\(P v) -> P $ v + fmap fromIntegral relPos)
              _ -> return ()
          _ -> return ()

makeUIState :: UIState
makeUIState =
  UIState
    { _bodyWidget = SomeWidget bodyWidget',
      _focus = []
    }

newtype Model = Model Int deriving (Show)

mkp :: Int -> Int -> BasePositon
mkp x y = P $ V2 x y

modelWidget :: [Int] -> Widget Model
modelWidget path =
  Widget
    { _width = 100,
      _heigh = 100,
      _model = Model 1,
      _backgroundColor = 30,
      _frontColor = V4 0 0 0 255,
      _visible = True,
      _path = path,
      _children =
        [ (mkp 15 25, SomeWidget $ textWidget [1, 2]),
          (mkp 15 50, SomeWidget $ textWidget [1, 2])
        ]
    }

instance WidgetRender Model where
  renderSelf bp w@Widget {..} = do
    renderer <- asks _renderer
    font <- asks _font
    liftIO $ do
      renderFont font renderer (pack $ show _model ++ show _path) (fmap fromIntegral bp) _frontColor

instance WidgetHandler Model where
  handler e a = do
    return ()

textWidget :: [Int] -> Widget Text
textWidget path =
  Widget
    { _width = 80,
      _heigh = 30,
      _model = "welcome ",
      _backgroundColor = 30,
      _frontColor = V4 255 0 0 255,
      _visible = True,
      _path = path,
      _children = []
    }

instance WidgetRender Text where
  renderSelf bp w@Widget {..} = do
    renderer <- asks _renderer
    font <- asks _font
    liftIO $ do
      renderFont font renderer _model (fmap fromIntegral bp) _frontColor

instance WidgetHandler Text where
  handler e a = return ()

initGUI :: IO (Renderer, Font, Manager)
initGUI = do
  initializeAll
  SF.initialize
  window <-
    createWindow
      "resize"
      WindowConfig
        { windowBorder = True,
          windowHighDPI = False,
          windowInputGrabbed = False,
          windowMode = Windowed,
          windowGraphicsContext = NoGraphicsContext,
          windowPosition = Wherever,
          windowResizable = True,
          windowInitialSize = V2 800 600,
          windowVisible = True
        }
  renderer <- createRenderer window (-1) defaultRenderer
  addEventWatch $ \ev ->
    case eventPayload ev of
      WindowSizeChangedEvent sizeChangeData ->
        putStrLn $ "eventWatch windowSizeChanged: " ++ show sizeChangeData
      _ -> return ()
  fm <- SDL.Framerate.manager
  SDL.Framerate.set fm 30
  font <- load "/usr/share/fonts/truetype/ubuntu/UbuntuMono-R.ttf" 20
  return (renderer, font, fm)

appLoop1 :: forall sig m. (UI sig m, MonadIO m) => m ()
appLoop1 = go
  where
    go = do
      e <- liftIO pollEvents
      -- SomeWidget bodyW <- gets _bodyWidget
      use bodyWidget >>= handlerSomeWidget e

      -- TODO: dispatch event to focus widget

      use bodyWidget >>= renderSomeWidget 0

      renderer <- asks _renderer
      present renderer
      man <- asks _manager
      delay_ man
      go

main :: IO ()
main = do
  (r, f, m) <- initGUI
  runReader (UIEnv r f m undefined) $ runState makeUIState appLoop1
  return ()